docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,67 @@
+# ADR 016: Validator consensus key rotation
+
+## Changelog
+
+- 23-10-2019: initial draft
+
+## Context
+
+Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
+key management policy.(https://github.com/tendermint/tendermint/issues/1136) So, we suggest one of the simplest form of
+validator consensus key rotation implementation mostly onto Cosmos-SDK.
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction(RotateValConsensusKey) that the new consensus key is now coupled with the validator operator with signature from validator wallet key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+
+### Considerations
+
+- consensus key mapping information management strategy
+    - stores history of each key mapping changes in kvstore.
+    - gaiad can search corresponding consensus key paired with given validator operator for any arbitrary height in recent unbonding period.
+    - gaiad does not need any historical mapping information which is past more than unbonding period.
+- limits
+    - a validator cannot rotate its consensus key more than N time for any unbonding period, to prevent spams.
+    - a validator should contribute X atoms to community fund to rotate its consensus key, also to prevent spams.
+    - parameters can be decided by governance and stored in genesis file.
+- slash module
+    - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.
+- data pruning
+    - database does not need to keep the historical key mapping changes after unbonding period past.
+    - when pruning is on, a node prunes all historical key mapping changes past unbonding period.
+- further implementation
```

Comment:
I would leave this out of the ADR and initial implementation for now.

Reply:
OK. I will remove data pruning and further implementation part.